projecttacoma/deqm-test-server
# Submit data

# Summary
We can now use custom operations to perform submit-data requests to the server

## New behavior
The server can interpret a FHIR Parameters object passed in using a POST request to http://localhost:3000/4_0_0/Measure/$submit-data. The server will parse the parameters and add new documents to the database for each contained FHIR object.

## Code changes
 - Added a transactionBundle Class to provide easy conversion from FHIR parameters -> transactionBundle -> JSON
 - Added new custom submitData operation for Measure type in config
 - Added submitData function which parses Parameters object into a transaction bundle and calls uploadTransactionBundle

# Testing guidance
- Start up mongodb using `brew services start mongodb-community@5.0` then run `npm run db:reset` and finally, `npm start`
- Open Robomongo and ensure it is hooked into the live database
- Create a valid FHIR parameters object, or use an existing one. I used the object found on [this page](https://www.hl7.org/fhir/measure-operation-submit-data.html) 
- Paste the object into the body of a request using Insomnia. 
- Submit the body in a POST request to `http://localhost:3000/4_0_0/Measure/$submit-data`
- You should get a 200 ok response with a list of entries and locations to find the created resources
- Use Robomongo to ensure that all resources in the parameters have been created
- Submit GET requests to `http://localhost:3000/{location url} for each response location url and ensure all created resources are GETtable
- Update and edit the Parameters object to try to break the upload, and ensure the server's response is appropriate
